### ISSUE TITLE
GitHub Contribution Guidelines

### ISSUE THREAD

**Issue by dblodgett-usgs** [cf-convention/cf-conventions#130]
Dear All,

As a next step toward the CF community using GitHub tools to discuss and refine the specification, we need contribution guidelines for this repository. For background and interesting reading, this issue follows #106 and #112 and is more or less governed by the CF community conversation in [Trac ticket 160.](https://cf-trac.llnl.gov/trac/ticket/160) The consensus in that ticket has lead us here.

The outcome of this issue should be an initial [CONTRIBUTING.md](https://help.github.com/articles/setting-guidelines-for-repository-contributors/) and modifications to the [pull request template](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/) per the requirements described below. We should use this issue to discuss and evolve these requirements then I will submit a pull request with content according to the consensus.

__Please review the use case and guidelines below. I will begin work on the CONTRIBUTIONS.md document in a few weeks based on the consensus we have sometime in May.__

__Top level use case:__ As a contributor to the CF specification, I need to know the rules and instructions for *how* GitHub should be used, so I know the right way to submit my suggested addition or changes. 

Guidelines: (derived from correspondence between @JonathanGregory, @dblodgett-usgs and others.)

- Some simple instructions are needed so that anyone who does not use GitHub for any other purpose can follow them successfully. GitHub can be used in a variety of ways and we need clear guidance regarding how the CF community wants to use the tools.

- A given proposal should be discussed as one issue. It shouldn't fork or be superseded by another one, unless that reflects what has happened to the proposal, in the same way that we continue discussion under one trac ticket for a given proposal. This is so that it's easy to trace the discussion which led to a given agreed proposal.

- It has to be easy to see what is being proposed in the context of the
discussion. In Trac we say:

   I propose to do X, Y and Z, with the result that the text will read as follows
   in Section N
   _blah blah blah_
   and as follows in Section M
   _rhubarb rhubarb_

   Thus it's possible to see the reasoning and the effect on all the sections, all gathered together, and understand it. In GitHub, there is no relation between the argument and the modified text, unless the argument is put as annotations in the text, but in that case the bits of argument are all over the place. **Somehow you need to get an overview of the changes all together.**   

- In the case of short changes, I think that excerpting the existing spec in a GitHub issue with suggested new content in the body of a GitHub issue would work well. Once agreed to in the issue, the change can be implemented and submitted as a pull request referencing the issue.

- In the case of long changes (i.e. where the body of an issue isn't long enough or changes are in many parts of the specification) the significant addition would likely come in the form of added content in a persons fork of the specification. They might register the contribution and their intent with the community in an issue for general feedback, work on the content in their fork, then submit a pull request when the consensus is that it's ready to be submitted. If the community finds utility in line by line comments in a pull request, that's great, but I don't think we need to force that issue. Line by line comments are really great for code. They can be a great for text, but they can also be very cumbersome. 

- Regarding "long changes" Maybe we could recommend that the discussion *begins* with proposed changes in the issue itself, but if the proposal involves changes in many places in the document, so it's awkward to spell them all out, it would be better to do this in a fork, but it's not worth doing that until it seems that the proposal is likely to be agreed. Moreover, if it does look like a proposal is going to be agreed, or when it has been, it would definitely be useful for the proposer to fork and update the documents, because otherwise someone else will have to do it later (as for trac tickets). So we should encourage that. However, to enable it, we need really simple instructions about how to do it!

- Regarding a hybrid approach of GitHub and trac, it seems that part of that ship sailed when we started managing the text in GitHub. That said, I see no reason not to allow changes to be vetted in trac until we have enough success using GitHub to track modifications that the community is ready to leave trac behind. We are already referencing trac from GitHub. Opening the option to not use Trac seams like the next step to me.

Best, 

Dave Blodgett

**Comment by JonathanGregory** [comment]
Thanks for starting this, Dave. I'm grateful to you for taking it on - it think it will be very useful. Since you've already included all my comments (!) I don't have any more to make just now. I hope others do. Jonathan

**Comment by ChrisBarker-NOAA** [comment]
Some thoughts:

Really great to get this out -- thanks for starting it!

Frankly, it's been a while since I've used TRAC much, but I think gitHub's Pull Request system provides some great opportunities to better manage changes to a document. IN the above, I see ideas like "Once agreed to in the issue, the change can be implemented and submitted as a pull request". But I think it makes more sense to go early with a PR. Often, the text itself what needs to be discussed -- so having the discussion directly in context of the suggested changes to the text has real benefits.

PRs support general discussion of the PR, as well as line-by-line comments specific text. This combination is ideal for many suggested changes to the doc.

So I suggest something like this:

### Issues:

If there is a general issue about the standard that isn't specific to text, like "I suggest we support his new concept in CF" then that should be in an issue.

If someone notices and error, inconsistency, missing detail, etc, but does not have a suggested change to the text, that should be an issue.

### Pull Requests:

In short -- go to a PR earlier rather than later -- as soon as someone has specific suggestions for changes to the text itself, that should go in a PR.

Any typo or minor correction should go straight to a PR. 

PRs and issues can be pretty easily linked, so the transition is not bad -- if someone submits a PR that requires a larger discussion unlinked to any particular text -- it's not hard to start an issue and link to it. Similarly, if it starts as a issue, it's easy to link to a PR once one is created.

## Potential Problems / Questions

### Making it easy for newbies

Creating an issue is significantly less of a technical burden than a PR. That is, pretty much anyone with no experience with gitHub or text-based markup can probably figure out how to submit an issue. PRs, on the other hand, require more tech know-how -- gitHub makes it pretty easy to automatically create a fork for you if you try to edit a repo you don't have access to, but it's still a bigger lift to wrap your head around.  So people should be encouraged to submit an issue if they have something to contribute (like correcting a typo, or...) that maybe *should* be a PR. A more experienced developer can make the PR later.

### What do "core" contributors do?

People without write permissions to the repo need to create a fork to make a PR. But folks with write permissions to the repo have three choices:

1) make a change directly and push it.
2) make a branch, make the change in the branch, and submit a PR for the change
3) make a fork into their personal account, and make a change and PR from there.

I personally prefer (2) as a core contributor to projects.

(1) should only be used for the simplest of changes like typos, etc.

Some folks prefer (3) as a more robust way to control the process, but It makes more work, and we're all volunteers here, so we want it to be as easy as possible to improve the doc.

In particular, (2) makes it much easier for multiple people to work together on a change -- rather than making PRs against forks againsts forks....

Just my thoughts, having worked on multiple gitHub projects, some large, and some small, some as core contributor, some not.

**Comment by graybeal** [comment]
Thank you Dave!

Well captured Chris, I concur. (And even for people who are newbies this could be a nice gentle introduction to Pull Requests!) 

I may be misremembering, but I thought that given the right configuration, text files in GitHub can actually be 'edited in line' and submitted as a pull request. (I do know that if you have enough privileges you can edit in line and have the results merged straightaway, but for sure that's not what we're doing in this case.) That could be a nice path to describe for people, if true.

**Comment by dblodgett-usgs** [comment]
Dear CF community, 

I've started work on #130. It can be seen in pull request form between branches in my fork [here](https://github.com/dblodgett-usgs/cf-conventions/pull/10/files) or in rendered markdown form [here](https://github.com/dblodgett-usgs/cf-conventions/blob/CONTRIBUTING/.github/CONTRIBUTING.md) 

@ChrisBarker-NOAA -- I did not get all your ideas implemented. If you want to leave comments with proposed additional details or ideas for an additional section, I'm happy to copy/paste/rewrite, I just didn't really know how to fit your thoughts into the doc at the first pass.

Others, please don't hesitate to leave in-line comments on the pull request version, but please do also summarize your comments in line in this issue so they make it to the email list.

Note that I also added the [Contributors Covenant](https://github.com/dblodgett-usgs/cf-conventions/blob/CONTRIBUTING/CODE_OF_CONDUCT.md) as a placeholder for a code of conduct. Is there a pre-existing CF code of conduct written down some where?

@JonathanGregory -- do you want to act as moderator of this discussion? I think it would be in our interest to move it to near conclusion prior to the meeting in June. 

Best regards,

Dave

p.s. apologies for the push of a couple commits to #115 that came over the email list a moment ago -- Those commits were promptly reverted and moved to another branch.

**Comment by ChrisBarker-NOAA** [comment]
@dblodgett-usgs wrote (in https://github.com/dblodgett-usgs/cf-conventions/pull/10/files):

"""
My opinion is that Pull requests are really not a good format to evolve text like this. They are great for critiquing code and doing fairly tight reviews on style and syntax, but for long form review of text, they can be very problematic.
"""

My experience, like yours, is that PRs are great for small changes to text, fixing typos, syntax rewrites, etc.

And yes, not as good for long form review of text -- but I fail to see how an Issue is any better for long form review of text -- in fact, it's much worse -- and in particular, if you not adding a whole new couple pages, but rather one small section, to a doc (or a couple small sections that are related but in different parts of the doc), then you get as much context as you want, without an extra typing or management. 

Now, for developing the first version of a largish doc with a small group, maybe Google docs is a better way to go, sure -- but I can't see how using Issues helps as all.

So again:

Once you are to the state of proposing a change to the text of the doc, whether it's a whole new section or a typo fix -- put it in a PR.

Also -- this is the first draft of the contributors doc -- whatever we do can be changed if things don't seem to be working as well as they might.

**Comment by dblodgett-usgs** [comment]
Dear Chris, 

Thanks for humoring me and working through this in the main repository issue. I commented back inline in my PR between branches in my personal fork of the repository, but will leave a note here as well.

**Why issues for discussion and PRs for syntax review?** Partly because @JonathanGregory requested that:

> A given proposal should be discussed as one issue. It shouldn't fork or be superseded by another one, unless that reflects what has happened to the proposal, in the same way that we continue discussion under one trac ticket for a given proposal. This is so that it's easy to trace the discussion which led to a given agreed proposal.

And because I totally agree. A proposal should be described as a problem or deficiency in the current spec (an issue). Then the solution planned and discussed among the community as one conversation that is traceable in one place. I think the issue interface is very well suited to this for it's markdown capabilities and integration with email notifications, among other things. The ability to link an issue to a commit and/or PR that resolves it is a very nice way to trace the connection between a problem, discussion of the fix, and the actual implementation of the fix. 

**Why a PR over in my personal fork and not here?** Two reasons, 1) so people can opt into the early line by line review / not inundate the cf email list with PR review spam and 2) to demonstrate what the interface is like without implying that we use it whole-hog. Per Jonathan and my observations, the PR interface is really not suited to commenting on sentences and paragraphs of text. It breaks the text apart and can quickly get out of hand. Look at how commenting works in google docs and word, that is much better suited to editing written word.

I hope that in the future, we have a nice interface for diffing, commenting, and evolving text for this kind of application, but right now we just don't. So we need to kick out to google docs or fork repositories in small groups to evolve the text of major contributions and look to the future for better solutions.

---
Rereading your and my comments, I think we are largely in agreement. I have a few minor changes I'll check into my fork in a moment. Maybe you could take my ["CONTRIBUTING"](https://github.com/dblodgett-usgs/cf-conventions/tree/CONTRIBUTING) branch and evolve the text to better convey the imperfect but pragmatic position we find ourselves in?

Best, 

- Dave

**Comment by JonathanGregory** [comment]
Dear Dave

Thanks for your contributions. I could be moderator of this, and thank you for
asking, but if David Hassell has time to do so that might be better since I'm 
rather short of time at present. Would you, David?

I agree with you that a googledoc is a good medium for editing large amounts of
text (in the mode where edits are suggestions). Maybe some comments could be
added about this possibility. There is a need to keep intermediate versions, I
would say, as the issue proceeds to discuss them. If there is only one evolving
doc it's not so easy to compare successive or alternative versions.

We should preserve the distinction between enhancements and defects that we
currently have with the trac tickets. Perhaps pure typos could be identified
as a separate category from these.

CF doesn't currently have a code of conduct but I agree it's a good idea to
adopt one. That's a governance issue, I think. I wonder what Karl feels.

Best wishes

Jonathan

**Comment by davidhassell** [comment]
I would be very happy to act as moderator, if that OK with you, Dave?

Adding a label to an issue is easy and should go into the [guidelines](https://github.com/dblodgett-usgs/cf-conventions/blob/CONTRIBUTING/.github/CONTRIBUTING.md).

I agree with what has been said about googledocs, but we should remember that they are not, at present, universally accessible and so we should always have alternatives.

David

**Comment by ChrisBarker-NOAA** [comment]
I agree with what has been said about googledocs, but we should remember
that they are not, at present, universally accessible and so we should
always have alternatives.

A google doc is also essentially ephemeral— it is not a good place to
record history, or the ability to roll back changes, etc.

We should use git+gitHub for what it is good for.

Granted, these are tools designed for code, not text documents, but when
you are working with a plain text markup language, they really do work
well.

In fact, being able to use them with a version control system is one of the
primary motivations to using text-based markup.

The CF standard is a pretty mature document. Most of the changes will be
small, and well suited to the “diff” based approach of gitHub.

So we should use gitHub for those changes.

Occasionally we may have a large new feature ( like the recent geometries
one ). If the primary authors of such a section find it more productive to
develop the first draft in google docs (or whatever they want), fine. But
it should go into a PR once there is a draft proposal.

-CHB

**Comment by dblodgett-usgs** [comment]
Dear all,

Apologies for going mute for a bit. Apparently, my email overlords (bow to the security gods) decided that github notifications are spam -- amazing how you don't miss notifications when they are gone.

I 100% agree that an external document is an imperfect solution that is really only suited to pre-community-review draft development. 

I also agree that we should have issue ["labels"](https://github.com/cf-convention/cf-conventions/labels) to `enhancement`, `defect`, and `typo`.  I will add a note that labels should be used if at all possible to the guidelines.

Yes, @davidhassell -- sounds great to have you moderate.

**Comment by erget** [comment]
I agree with the discussion wholeheartedly, which for the sake of keeping things tidy and confirming my own understanding I summarize as follows:
- If you want to change something and are in doubt, raise an issue
- For anything else, a PR is a good way to go, although an issue isn't harmful, especially if it's about a typographical matter
- What I'm missing is a release strategy, which may need to be addressed elsewhere. I propose something inspired by Git Flow, roughly as follows:
  - `master` is the current release.
  - `next` is the next release, an "editor's draft", if you will.
  - If you want to make a change, make a branch based on `next` and implement it. Then send a PR to merge it back into `next` or ask for help.
  - When a change has been approved, it's merged into `next`.
  - When we make a release, we merge `next` into `master` and tag that.

Perhaps more refinement is possible - we discussed at the meeting having some kind of a beta version of a next release for a time. This would be compatible with this workflow (which is one of many possible ones, but fairly simple and IMHO straightforward).

**Comment by cameronsmith1** [comment]
This seems reasonable to me.  I would expect that some active management may be needed in the run-up to a new release, ie managing which PRs get merged or held until after the release.  However, that shouldn't affect your general outline.

**Comment by ajelenak-thg** [comment]
>I propose something inspired by Git Flow, roughly as follows:
>master is the current release.
>next is the next release, an "editor's draft", if you will.

I don't think we need a `next` branch. Each new convention version will be a tagged release in the master branch thus allowing this branch to represent the Editor's Draft. All the new material merged into the master branch after the last release is assumed to be accepted for the next release, barring any stylistic or formatting changes.

The process would go something like this:

1. Anyone who wants to propose a change to the convention should fork this repository. (I assume the number of those with commit privileges will be small so forking is what majority will have to do.)

1. Create a branch off the master in the forked repository and work on the proposed changes in this new branch.

1. Creating a pull request to the upstream (the official convention's) repository initiates the formal review process of the proposed changes.

1. Pull request creator is responsible for updating its branch and the text of the changes with the upstream's master branch during the review process.

1. The review process ends when the pull request branch is merged into the upstream's `master` branch.

It would be good to appoint, at least nominally, a few editors for every new convention release. They would be in charge of preparing the text in the master branch for the next release and making sure the review process of any open pull request is reasonably timely.

**Comment by dblodgett-usgs** [comment]
I 💯% agree with @ajelenak-thg. 

I think this is more or less how the repository works now and this is a very natural pattern.  Released tags would get built and stored as binaries here: https://github.com/cf-convention/cf-conventions/releases as well as on the main cf web page.

**Comment by erget** [comment]
This is definitely the most natural way to work on GitHub and very git-native, which I like. However, I do think there is merit in having a `next` branch. Because CF isn't software that we're releasing with continuous delivery, having fewer, but larger, changes between versions is probably the better way to go. It also better reflects the understanding achieved at the meeting - namely, that `master` is basically a release branch, and that the draft (whether named `next` or something else) collects changes before merging them all into `master`.

Concerning the workflow @ajelenak-thg outlines - that's certainly the way I intend to work, but my feeling is that we'll perhaps need to have a bit of patience with people who are new to both git and GitHub as they get used to it, so we shouldn't be too strict with enforcing the specifics of that workflow.

So the only point of ambiguity for me is the branches. I do think that having `master` only be merges with tagged releases has advantages for novices - even though you can achieve the same amount of traceability by only tagging releases, and declaring intermediate commits on `master` as transitionary, I think that having non-tagged releases appear as the repo's landing page could be confusing to some of our contributors.

**Comment by marqh** [comment]
I think that the people who are new to github will appreciate a lack of complication in the structure of information and how to engage.

Having a single branch makes is clear to anyone arriving that if they want to propose a change, that is what their change is with respect to.

So, I am in favour of maintaining the `master` as the editor's draft and providing tags for released versions and against a `next` branch.

I have also proposed a change to the CF website that makes the current state a little more clear: https://github.com/cf-convention/cf-convention.github.io/pull/61

**Comment by erget** [comment]
I don't feel strongly about how we do branches and since I'm the only one currently supporting having a `next` branch on here that's fine with me. I think the release frequency is something which should be discussed more widely, but it's independent of our workflow for producing releases.

**Comment by davidhassell** [comment]
I think that we have to be very clear this is a proposal for how github issues can replace Trac tickets (as opposed to anything else we currently use, like the mailing list) and, at the this time at least, nothing else. This issue is concerned with how best to use github to carry this out in a way that make sense to people now, and in 20 years time.

In particular, github issues are only for completely defined proposals, not for any sort of speculative change; and all proposals must be raised as an issue (although that could coincide with a pull request for simple cases, e.g. typos).

The discussion has widened a bit to include "how should we integrate accepted proposals into CF", which is fine, but this is a separate concern to the original "how should we get a proposal accepted". These two are certainly connected, but will need quite distinct guidelines.


The [original proposal](#130) contains guidelines that I think say

1.  Start by using a github issue just as you would a Trac ticket
2. ...
3.  When it is finished there will be a pull request

How we get from 1. to 3. is less clear to me, but that will be resolved by the discussions in this issue.

When I come back in 2038, I need to be able to start at the top of #130 and be able to follow the arguments that led to its conclusion. Linking out to pull requests as part of this read through is fine by me, although I have a concern about searching - if I search for "foobar" in the issue, will it also search for that string in all referenced pull requests? This was also an issue for documents attached to/linked form Trac tickets, but the reality was that they were very rarely used, and if they were it was not for essential items such as proposed text changes.

**Comment by JonathanGregory** [comment]
Dear all

As David H says, I think we should first decide the procedure for making decisions on changes using GitHub (as an alternative to Trac). This will be part of a modified version of http://cfconventions.org/rules.html, so at this point we need explicit text to put in that document. In principle it should be the same procedure as now regarding the time-limits etc.

One change in the procedure could be to recognise an extra kind of change request i.e. typo, which is different from defect or enhancement. A defect is a proposal to change the words of the document to correct an error or clarify them, with materially changing the meaning of it. Someone else might disagree with the proposer of a defect ticket and think what they propose is actually a material change to the meaning, and in that case the proposal has to be discussed as an enhancement instead i.e. not accepted by default. A typo is an even more minor change which fixes something that seems evidently to be simply a mistake. However it's possible, in an analogous way, that someone else might say that it's not a typo, but deliberate, in which case it'd have to be discussed as a defect or enhancement instead.

I think that a typo could be proposed with a pull request, but a defect or an enhancement should be started with an issue, as David H says, and proceed to a pull request when it's fairly well agreed and it comes to a matter of wording.

Cheers

Jonathan

**Comment by davidhassell** [comment]
I have a concern with the hybrid github/trac approach: Some github issue numbers are/will be the same as Trac ticket numbers. 

Is this managable?

David

**Comment by ajelenak-thg** [comment]
It is manageable only with additional context, for example: `Trac #101` or `GH #101`.

Alternatively, GitHub's API can be used to create and close dummy issues until the internal numbering is bumped up above the highest ticket number on Trac. Or go all the way to, say, 200. So all ticket numbers below 200 belong to Trac; above 200 to GitHub.

**Comment by ChrisBarker-NOAA** [comment]
So all ticket numbers below 200 belong to Trac; above 200 to GitHub.

+1

**Comment by sebvi** [comment]
why not simply import all past TRAC tickets and convert them into github issues? 
I am sure we can find that kind of scripts, possibly in a github repo!

**Comment by sebvi** [comment]
1mn googling: https://github.com/mavam/trac-hub

**Comment by dblodgett-usgs** [comment]
Dear All, 

I would be supportive of a clean break from Trac -- and it's clear that there is a constituency in the community that would prefer that path. I think it is also clear that we are suggesting that outcome in the near future anyways, but potentially not an immediate retirement of Trac. While it may introduce a bit of confusion, my opinion is that confusion may be worth while to keep all community members engaged and be as inclusive as possible. At the end of the day, those of us that prefer using github need to demonstrate how to evolve the standard in github so those without experience in this environment can learn and join the :shipit: tribe. (I use the [squirrel](http://excid3.com/blog/finishing-is-all-that-matters) as an example of how foreign some github traditions and mechanisms really are.)

Is there any strong disagreement to this line of argument? My thinking is that we would continue to allow use of Trac until usage has dropped to near nill or maintenance of Trac is not worth the effort. Putting a timeline on this may not be worth the energy as it should happen naturally if this switch to github really is the thing to do.

I also think it would be a great idea to migrate the Trac history to a block of github issues. While the numbering would reset, we could embed the original Trac URLs in the issues for continuity. I think this would naturally occur at the time that we retire our use of Trac.

Best, 

Dave

p.s. I would like to remind everyone that while we have moved some of our discourse to github, I think we should attempt to maintain the traditions of CF in our correspondence. Each of these comments fires off an email to the group and, as such, should be thoughtful and worthy of people's attention if at all possible.

**Comment by davidhassell** [comment]
OK - I prefer the second option, because I can see the prefix getting dropped somewhere down the line, and the format isn't really part of the proposal.

Most of the existing github issues record the process of creating 1.7 from the backlog of concluded tickets in Trac. The github issues seem like the natural place for these things and, for example, the recent ticket on an asciidoc issue (#135).

Perhaps we need a new, seperate repository for change proposal e.g. `cf-convention/change-proposals` (or some such name). Then we can keep issues in each of the separate repos (`cf-conventions`, `Conformance`, `cf-convention.github.io` for all the other practical (rather than scientific or logical) thing that crop up.

Do pull requests work across repos in the same project?

Thanks, David

**Comment by taylor13** [comment]
Hi all,
I am told that PCMDI's maintenance of the Trac site here at LLNL is becoming more fragile for a number of technical reasons and our scarce resources to invest in it also make it problematic, so from a practical perspective our systems folks here would favor moving to github (although it is not imperative that this happen immediately).

Like others I am concerned about consistency of and/or conflicts between issue numbers associated with the Trac and github discussions.  If we renumber the trac issues won't it disrupt the cross-referencing already in those discussions?
Karl

**Comment by dblodgett-usgs** [comment]
@davidhassell 

I think there may be some logic to creating a new repository that is dedicated to change proposals and community governance. That way people could "watch" that channel for topical conversation rather than syntactical review that needs to happen on the asciidoc. Anyone else want to chime in on this proposal? If we decide to go down that road, I will need to rework much of the content I've been working on for this, but that's all good -- just want to bring this to some form of conclusion.

Regarding integration of pull requests across repositories, look up in this thread a bit. You'll see a link from https://github.com/cf-convention/cf-convention.github.io/pull/62 that got put there when I linked to this issue from an issue in: https://github.com/cf-convention/cf-convention.github.io

Best, Dave

**Comment by erget** [comment]
I think there could be some benefits for this, but yes, we'd need to update the contribution guidelines to be really clear. Perhaps the easiest thing to do would be to clone the cf-conventions repo and rename that so that topical pull requests / discussions go there, and then once things have been integrated on that side pull them across to e.g. this repo, which would be used for releases, including syntactical / editorial changes in order to keep the adocs up to date. As @dblodgett-usgs notes this would be easier for users wanting to only be involved in the content-related discussions.

For me this seems overly complex, but if the community really wants to be able to restrict spamming it could be a good way to go. Personally, I think the better alternative would be to post to different lists based on issue tags - I've heard of Hubot integrations which allow this, for example.

In that vein, I'd also be in favour of moving the Conformance into this repo, because that would make the workflow easier for ensuring that both the Conventions are updated simultaneously (I'm thinking of PR checks, as an example).

**Comment by rhattersley** [comment]
As a general observation on the transition from Trac to Github, I would encourage embracing existing GitHub idioms to take advantage of the resulting simplicity. For example:

re: porting Trac issues to GitHub, I would recommend that you copy them to a separate repo dedicated to being an archive. That way the issue numbers can easily be preserved. If there's a need for new native GitHub issues to refer back to the Trac content, cross-repo links are still very easy, e.g. cf-convention/test-trac-convert#1.

re: conversations on "practical" vs "science" topis, I would recommend keeping all your issues and pull requests in the repo that they affect. Workflows across multiple repos are more complicated and error prone, and much less likely to receive any attention from GitHub. @erget's suggestion to take advantage of issue labels is a more natural fit. The use of issue labels is very common on GitHub.

**Comment by JonathanGregory** [comment]
Dear all

I agree with @rhattersley that it would make sense to copy the trac tickets to a separate cf-conventions GitHub repository, with their original numbers, in order to keep them for posterity. I don't think we should oblige everyone to switch immediately to GitHub and stop using Trac, though we can encourage it. The meeting last week was certainly in favour of a quick move, but the people at that meeting may not be an entirely representive sample of the whole community of contributors to CF. I guess it was a more technically minded set than the average.

We currently have three main repositories: the website, the conventions and the conformance document. I suggest that, for simplicity and unless it becomes inadequate, we could deal with governance issues in the website repository, because the gobernance rules are part of the website. This ticket ought to be in that repository, not this one, if you agree with this idea. I agree with @erget that it would make sense to merge the conformance and conventions repositories, since they have to be updated together. The data model will also belong with them.

I don't know how the mechanism works for sending these postings on issue to the CF email list. If this system can filter by label, we could follow what @erget and @rhattersley suggest to mark some of them as not to be distributed.

Best wishes

Jonathan

**Comment by ChrisBarker-NOAA** [comment]
On Thu, Jun 28, 2018 at 5:50 AM, JonathanGregory <notifications@github.com>
wrote:

> I agree with @rhattersley <https://github.com/rhattersley> that it would
> make sense to copy the trac tickets to a separate cf-conventions GitHub
> repository, with their original numbers, in order to keep them for
> posterity.
>
+1

> I don't think we should oblige everyone to switch immediately to GitHub
> and stop using Trac, though we can encourage it.
>
yes, we should -- having the same thing managed two ways is a pain in teh
#^$*.

I know change can be hard, but semi-change / lack of clarity, confusing
mess is harder.

And it's really not that different for the common use cases!

> The meeting last week was certainly in favor of a quick move, but the
> people at that meeting may not be an entirely representative sample of the
> whole community of contributors to CF. I guess it was a more technically
> minded set than the average.
>
maybe so -- but TRAC isn't exactly a model of usability :-) gitHub (at
least for issue management) isn't any more technical that TRAC, and I think
it's more usable, not less. So it's more a old dog / new tricks issue.

-CHB

-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by davidhassell** [comment]
A summary of where I think we are:

There seems to be a consensus for:
  * The proposed [guidelines](https://github.com/dblodgett-usgs/cf-conventions/blob/CONTRIBUTING/.github/CONTRIBUTING.md)
  * A separate repository for the existing Trac tickets to be copied to issues
  * The appropriate use of issue labels 

I think that some more discussion is required for (perhaps some of these questions are nearly agreed):
  * whether or not to run Trac in parallel for a while
  * whether or not to create new github issues in the cf-conventions repo, the same repo as the copied Trac tickets, or yet another repo. [Whatever the answer, labels will be in use to help discern issues.]
  * whether or not to move the conformance document to the cf-conventions repo

I'm assuming that the discussion on how/when to merge completed proposals will be carried on in a different discussion, so we can move the original proposal of #130 to a faster conclusion.

Thanks, David

**Comment by dblodgett-usgs** [comment]
I agree.

On the outstanding issues:
1. Trac in parallel? I think we are already in a phase where it is being used in parallel. The goal should be to shut it down, but as I've argued in the past, it may be too much to require a sharp transition. 
2. I really think we should try using one repository and the email list. The cleanest solution would seem to be to migrate trac tickets to a new repository and bring the git commit history from https://github.com/cf-convention/cf-conventions into it. From that point forward, https://github.com/cf-convention/cf-conventions would be deprecated and all new work would take place in the new repository. I think that naming shouldn't be a problem. The new repo could be `cf-convention` or `cfconventions` or maybe we could just swap the names?
3. I would like to suggest that the conformance document be a separate issue all together.

**Comment by dblodgett-usgs** [comment]
@erget 

The benefit of making a new repository would be the ability to migrate trac history 1:1 preserving issue numbers. I don't feel strongly about this at the end of the day though. My only strong opinion is that we not split discussion into a repository for discussion only. I think we should try to keep it as tight as we can.

Yes, I think the conformance document should be treated as a separate discussion from the Trac to Github migration. Clearly related, but not enough to muddy this discussion with that.

Best  -- Dave

**Comment by davidhassell** [comment]
Where I think we are on the outstanding questions:

  * **whether or not to run Trac in parallel for a while**

    I suggest that, for new proposals, we do not run Trac and gihub for new proposal parallel. In practice this would mean:

    1) We would set a date (that would presumably coincide with the acceptance of this ticket) after which new proposals must be made on github and no new proposals will be allowed on Trac. Discussions can continue on Trac until ...
    2) We set another, later date, after which Trac will be turned off and its content archived. Any unfinished Trac discussions will be effectively closed. If there is still a will to continue a discussion then the proposal must be raised anew as a github issue proposal. (This is not as extreme as it first sounded to me - there are Trac tickets that have been open but haven't been posted to for over 10 years - these were never going to be resolved before Trac is turned off!)

  * **whether or not to create new github issues in the cf-conventions repo, the same repo as the copied Trac tickets, or yet another repo. [Whatever the answer, labels will be in use to help discern issues.]**

    It seems that there is support for copying the existing Trac tickets to a new "archive" repository; but raising new proposals in the existing cf-conventions repository. New proposals can happily coexist with other issues (such as "should we use italics for example captions?") with the use of labels.

    We would "fast-forward" the issue numbers in the cf-conventions repo so that new proposals would have different numbers to old ones - currently that means creating dummy issues up to and including nunber 174. This is important because if we didn't do this then our issue referencing could break in the future if we were to alter our change procedure yet again (such as if github were to ever fail to meet our needs). 

  * **whether or not to move the conformance document to the cf-conventions repo**

    There is a compelling reason for moving the conformance document to the cf-conventions repository in that then there would only ever need to be one pull request per completed proposal. In the current situation (i.e. conformance in its own repo), some proposals will need two pull requests.

Thanks, David

**Comment by graybeal** [comment]
These days I am not a heavy user, so please discount these observations as you see fit.

If I am not mistaken, David H's bullet 1 is affected by bullet 2. Namely, *if* you are going to copy from Trac to GitHub, and *if* wherever that copy gets made is used to continue discussions on those points, then all Trac activities cease as of the change-over day, to be replaced by GitHub. Which I think would be most time- and cost-effective.

This next may sound radical but as a practical matter, if it is only 174 tickets, this need not be a huge deal. If there are people spending 50% of their time reading Trac tickets, then maybe it needs to be done just right. But for 174 tickets, many of them closed, I think even if you don't make the numbers track with trac (sorry), the total number of impacts is going to be small on day 1, and vanishingly small after a year. OK, 2 years.

1) Step one is copying the tickets. As mentioned, there are at least 4 GitHub projects and this Slack page lists a number of approaches, some of which look quite mature:  https://stackoverflow.com/questions/6671584/how-to-export-trac-to-github-issues. Copy the tickets at the end of the current repository (existing ticket #$), making sure Trac's closed tickets are closed in GitHub. (I'm not sure you'll be able to save the dates of all the entries with these tools; but if it's easy, you could copy the dates as the first line of each entry/comment. Copying the original Trac # wouldn't hurt either.)
1A) If you absolutely had to clean up all those links, somewhere in the transition pipeline write a script to convert all the links in those 174 tickets to point to the new GitHub ticket/comment location, if that isn't handled automagically by the conversion tool.
2) Either manually or via a script, create an entry for the last comment of each existing Trac ticket that says  "Discussion and maintenance of this issue has been moved to <new_uri_for_old_trac_ticket_n>."  
3) If you feel compelled to have a forward reference after the Trac repository goes entirely away, use a separate repository to copy the 174 tickets (preserving the ticket number) with just the title of the ticket, and an entry that says the same thing as above. Or create a ticket-mapping entry.  
   But if you count the number of open Trac tickets, the number of comments being actively made on them, the number of people making them, and the number of cross references those people need to follow, I propose it is not going to be a major crisis for everyone to adjust to the new location in the new repository, especially given the titles are the same and the previous Trac (or a copy of it) has a forward lookup/reference. 

I don't think the fact that all of this could change again someday is an argument that says the new numbers have to match the old. Either way all the numbers/references might have to change again. 

 It would definitely be ideal if numbers were unique and matched; but even if some were duplicated and they didn't match, you will always be able to manually recognize that for all the new github tickets that came from trac (from #$+1 to #$+174 in the github repository), the internal numbers will need to be mentally cross-referenced into the new Github numbers (original number minus #$). This could be done automatically at the time of the next conversion, if it isn't done this time around, or even manually as desired on a per ticket and per comment basis.

**Comment by dblodgett-usgs** [comment]
Hi All,

This conversation has come quite a long ways. Glad to see that we are coming to some conclusions. 

I've done a bit of cleanup on the "CONTRIBUTING.md" document in the pull request mentioned above. Please comment and edit away. 

If we can work that text to finality, I think we just need to pick the timing of the switch over and put together the task list to get us there. Lots of food for thought about a task list above.

Best, 

Dave

**Comment by erget** [comment]
One thing we might wish to include in the contribution guidelines would be the branch which contributors should contribute to. As it is we say they should create an issue and a pull request which matches that issue, but there are some branch management issues which would have minor effects on this. If we have a branch for the next version of the convention, say `1.8`, should everybody base their branches off of there? Otherwise, if they're always branching off of `master`, there's the possibility that they'll fiddle with areas which will produce conflicts later, e.g. the author list or the version number.

Alternatively, we could say that the files which contain such common metadata related to the convention itself shouldn't be edited in a topical PR, but rather only be changed editorially in preparation for publishing a new version.

**Comment by dblodgett-usgs** [comment]
According to the discussion above, I think we've agreed that developments beyond the most recent [tagged and deployed version](http://cfconventions.org/documents.html) would take place on master.

I 100% agree that we should consider updating "history" and other common metadata as you say when we cut a release to avoid merge conflicts. I'll add something to this affect unless there is objection?

**Comment by JonathanGregory** [comment]
Dear Dave

Thanks very much for moving this forward. I think it's almost there. When agreed, this will be a change to the governance rules on the website, so it really belongs in that repository, but never mind!

* General guidelines 1. "This is so it is easy to trace the discussion what" should be "that".

* General guidelines 3. Would it be OK to state (in view of 1) that there should always be an issue for a defect or an enhancement?

* General guidelines 4. "Please attempt to present contributions as enhancements, defects, or typos." Can't we insist on this? Are there other possibilities? Different rules apply, so one must be clear which kind is intended.

* Typo fix. No need for an issue because these won't be mentioned in the history. That might be worth saying.

* Single section change. "If the modification is non-controversial (especially likely for a defect issue)".

* Multiple section. "should follow the patter". -n needed. Could add to this point that rich-diff is a very useful way to see what's proposed to be changed!

Best wishes

Jonathan

**Comment by dblodgett-usgs** [comment]
Dear Jonathan, 

Actually, this text will end up in a "CONTRIBUTING.md" in this repository. It may also end up on the main web site. [This pull request ](https://github.com/cf-convention/cf-convention.github.io/pull/62) likely needs some attention in light of the conversation here.

- Changed which to that.
- Yes, I think we should expect issues to be closed by practically all pull requests. I've modified the text.
- Yes, we should insist that those labels be used for **contributions.** Note that we may have labels for questions or other kinds of "issues".
- added: "as these changes do not appear in the convention history."
- Fixed type.
- Added "Note that their is a rendered "rich-diff" view of a pull request that can be helpful for review of large contributions."

I will get these changes checked in soon.

Best, 

- Dave

**Comment by dblodgett-usgs** [comment]
Dear All,

I got the changes noted in my previous comment implemented.

I've also updated the rules to reflect our switch to github only. Some cross linking of the github CONTRIBUTING.md and the rules are in order? 
- rules PR: https://github.com/cf-convention/cf-convention.github.io/pull/62/files?short_path=91508a7#diff-91508a78c1f574c8c31da226cf3e7622
- Rendered Rules: http://cfconventions.org/rules.html
- Latest CONTRIBUTING.md PR: https://github.com/cf-convention/cf-conventions/pull/137/files?short_path=8cc7b2b#diff-8cc7b2b0d78dd2501610391c086a8516

So what are our next steps? It seems we have come to a basic consensus and should pursue a role out of this strategy?

Best, - Dave

**Comment by JonathanGregory** [comment]
Dear Dave

I think you've done an very good job of this. Thanks a lot. I have a couple more points:

* There are separate rules for defects at http://cfconventions.org/errors.html.  I think these need updating too. In the paragraph beginning "If someone thinks"
we perhaps now need to distinguish "defect" and "bug".  

* I would argue that we need only the bug, defect and enhancement labels at https://github.com/cf-convention/cf-conventions/labels. An asciidoctor mod is a bug, I'd say, being typographical. I don't know what "duplicate" is. The others ("help wanted", "invalid", "question") are matters to be raised on the email list, not GitHub. I believe the plan is just to replace Trac with GitHub for managing the document, not to provide an alternative venue to the email list for discussing CF - right? In point 4 of the general guidelines you mention "typos", not "bugs".

David is moderating this discussion, I believe. If we're following the usual rules, he should summarise the state of the discussion, and if no-one says anything more for three weeks, it's a done deal. However, David is on holiday this week, and I think next week. Sorry about that.

Best wishes

Jonathan

**Comment by dblodgett-usgs** [comment]
Dear Jonathan,

I've updated https://github.com/cf-convention/cf-convention.github.io/pull/62 to include changes to errors.md.

I opened https://github.com/cf-convention/cf-conventions/issues/131 to deal with labels. My suggested changes are:  
Remove: `asciidoctor mod?`, `bug`, `invalid`, `simple`. 
Add: `typo`, `style`

OK, I'll wait for @davidhassell to summarize.

- Dave

**Comment by davidhassell** [comment]
Hello,

I'm back from leave. This is where I think we are with this discussion.

  * "Rules for correcting errors in CF documents" **[1]** and "Rules for CF Conventions Changes" **[2]** look good to me and have no outstanding comments.

  * "Contributing to the NetCDF-CF conventions" **[3]** is also looking good, but I wonder if we need to make note in this document of the issue that @erget raised: which branch should pull requests be created for?

    I favour creating a branch for the next version (e.g. `1.8`) and specifying that pull requests be made to this branch. This branch would be created as a copy of the previous version (1.7) when that is released. An advantage of this over using `master` is that accepted issues can be merged into the branch prior to release, and so be picked up by PRs for other, newer issues. I don't have particularly strong views on this, though - on which branch nor whether this should be in the CONTRIBUTING.md document.

**[1]** https://github.com/cf-convention/cf-convention.github.io/pull/62/commits/e6f7747c8b94952651c567c9ffead9f74131080b?short_path=8f8a0b0#diff-8f8a0b09fa5f199438dd9e7557ce36db

**[2]** https://github.com/cf-convention/cf-convention.github.io/pull/62/files?short_path=91508a7#diff-91508a78c1f574c8c31da226cf3e7622

**[3]** https://github.com/cf-convention/cf-conventions/pull/137/files?short_path=8cc7b2b#diff-8cc7b2b0d78dd2501610391c086a8516

**Comment by marqh** [comment]
@davidhassell 

> An advantage of this over using master is that accepted issues can be merged into the branch prior to release, and so be picked up by PRs for other, newer issues.

Please can you elaborate on this statement?

It seems to me that it is fine to merge accepted issues onto `master`, that this should be the approach that is adopted.

What makes you make the statement that having an explicit branch for a next version allows merging accepted issues, suggesting that targeting `master` means that accepted changes would not be able to be merged?

thank you
mark

**Comment by ChrisBarker-NOAA** [comment]
> I favour creating a branch for the next version (e.g. 1.8) and specifying that pull requests be made to this branch. This branch would be created as a copy of the previous version (1.7) when that is released. 

The "usual" git process (for code) is that a "release" is tagged (and maybe in its own branch, whereas "master" is used as the active copy -- being constantly updated -- when it it ready for a new version, a release is made.

Kind of the point of "master" is that it is the default branch, that's seeing activer development.

Why use  a different workflow here??

**Comment by marqh** [comment]
i agree with @ChrisBarker-NOAA 


I am in favour of instructing contributors to propose all changes to `master` and tagging releases as they are created.

Branches are not required for managing the document in its normal work flow (though they may be useful for occasions such as  where an updated 'point release' is required to an existing release).
This is for administrators of the repository to manage, not for contributors to have to negotiate.

I think that managing branches for each 'next version' will cause maintenance problems and confusion for contributors.

There are already loads of branches within this repository, which can/will lead to further confusion, especially if we ask contributors to understand why these branches are there and which one to pick.

I recommend keeping the use of branches limited to repository administrators and pointing all contributions and contributors at `master`.

I also recommend ensuring that accepted changes are merged to master as soon as they are approved, so that further changes are being made with respect to the latest approved changes.

I am fairly clear on the benefits of this approach and concerned about the risks of pointing contributors at branches and holding off merging agreed changes to some future time.

**Comment by ChrisBarker-NOAA** [comment]
@hrajagers wrote:
>The branch per "release" approach might make it conceptually a little bit easier to start working on a draft of "release 2.0" while also working on an earlier "release 1.8".

Agreed -- master should be for the current "working version" -- i.e if 1.8 is out, changes that will be in 1.9 go in master.

In the usual case of software, you have a "current release", say 3.2 (choosing examples that aren't currently used for CF), and you are working on 3.3 in master.

Meanwhile, if you are maintaining older versions, say 3.1.3, you might do bug-fixes on that, and release a 3.1.4 -- in which case, there is a branch for 3.1.* where the bug fixes go.

In the CF case, we don't have to fix bugs in older releases, so it's easier.

The one complication is if we have a "future" version that wont' be released for a while -- that would be 2.0 for CF -- so that should be in a separate branch -- or maybe a separate repo altogether.

I jsut took a look at the cPython repo -- for an exampel of a major project:

https://github.com/python/cpython

The currently have:

2.7
3.4
3.5
3.6
3.7
master

3.7 is the latest version -- they are working on 3.8 in master, and 2.7--3.6 are in maintenance mode (i.e. just bug fixes).

**Comment by ChrisBarker-NOAA** [comment]
> There are already loads of branches within this repository, which can/will lead to further confusion, 

Those look like "feature branches" -- which can be "pruned" once they are no longer needed.

https://railsware.com/blog/2014/08/11/git-housekeeping-tutorial-clean-up-outdated-branches-in-local-and-remote-repositories/

Note that if a branch has been merged, then pruning that branch doesn't lose any information in the history.

**Comment by dblodgett-usgs** [comment]
I agree with @marqh but want to recognize that this should be, to some extent, at the discretion of the repository admins and issue moderators. I say this because a given change may have a different disposition with regard to the current document depending on the state the overall project is in at the time. e.g. we are working hard on v2.0 but v1.9 is about done. I wouldn't want to 100% pre-judge how people want to use branches for a given PR at that point in the future. 

So -- 
1. PRs get opened against master which should be even with or recognizably ahead the latest release.  
1. past releases are tags on master. No branch needed because we don't bug fix old releases.
1. branches are used as short and/or long term feature development space.
1. approved changes are merged to master as soon as possible after acceptance.

Then the question is, should we add something to this affect to [CONTRIBUTING.md](https://github.com/cf-convention/cf-conventions/pull/137/files?short_path=8cc7b2b#diff-8cc7b2b0d78dd2501610391c086a8516)?

**Comment by ChrisBarker-NOAA** [comment]
Sounds good -- and yes, this should go in CONTRIBUTING.md

**Comment by cameronsmith1** [comment]
We started using Github a few years ago for a project I work on. It was decided at the start that changes should go onto master frequently, and that we would simply tag master for the major release versions.   It didn't work in practice, because near release time there were divergent needs within the project.  Some people wanted to only allow changes to master that were for finalizing the release, while other people needed to keep working on future capabilities.  If finalizing the release only took a couple of weeks this wouldn't be a big deal, but big releases often takes months to finalize (CF tends to spend a lot of time agonizing over the details too).   I think this is the issue that @dblodgett-usgs mentioned in his last post (above).    However, I don't see it explicitly mentioned in the 4 bullets.

There are two ways I can see this working near release time: (a) master is for the release version and future changes go onto one or more feature branches that get merged to master after the release, or (b) vice versa, ie the release on a branch and the master keeps developing.  In my project we ended up doing (b), but it would probably have been better to do (a).

In any event, I think it would be good to add a fifth bullet to the list of @dblodgett-usgs that clarifies how master and branches will be used near a release.

**Comment by ChrisBarker-NOAA** [comment]
@cameronsmith1: good points.

I think the way to do it is to crate a "release candidate" branch when getting close.

-CHB

**Comment by dblodgett-usgs** [comment]
@cameronsmith1 -- my point on that was that I don't think any guidance we write now is going to be right. So we should use our best judgement on the workflow when we get there. I would rather not attempt to write rules for that stuff until we have more experience.

**Comment by davidhassell** [comment]
There are clearly good reasons for making PRs against master, so I withdraw my suggestion of a using a branch.

I like 1, 2 and 4 of Dave's list (https://github.com/cf-convention/cf-conventions/issues/130#issuecomment-412947016) but don't see the need for point 3: "branches are used as short and/or long term feature development space.". This, I think, is not relevant to creators of PRs so we don't need to mention it in this document.

The same seems true for the management of near release time - we don't need this in CONTRIBUTING.md because it won't affect contributors, who would carry on making PRs to master, regardless of the release mechanics being employed at the time.

**Comment by dblodgett-usgs** [comment]
Agreed - @davidhassell. My number 3 is more about how the sausage is made -- which we'll work out as things go forward.

I'll get the CONTRIBUTING.md updated soon.

**Comment by cameronsmith1** [comment]
Removing 3 from Dave's list certainly makes it clean and simple.   The other issues can then be deferred until they are needed, per long-standing CF tradition.

**Comment by ChrisBarker-NOAA** [comment]
"I am assuming here that 'someone who wants to create a pull request' may actually want to create a rather complex change and work on it a little while with other folks"

I do that that is a significatn use case -- but whether that is a feature branch or a PR from a fork depends on involved someone with permissions on the repo is.

If an "outsider" starts a PR against master, the repo managers can choose to merge that into a feature branch instead of master if they want to develop it for a while before merging.

In short -- creation of a feature branch is the job of the core maintainers, so I'm not sure it needs to be mentioned in Contributing.md

Though maybe should be mentioned somewhere else as part of recommended workflow.

**Comment by davidhassell** [comment]
I looks like [CONTRIBUTING.md](https://github.com/cf-convention/cf-conventions/blob/23c0db289c3422b2fbf5a618f987f836577b490d/.github/CONTRIBUTING.md) is in good shape and,  as I see it, there are no outstanding issues. If there are any further items to consider please post soon, as otherwise this issue will be accepted as is after the 3 week deadline is up. Thanks!

**Comment by cameronsmith1** [comment]
It looks good to me.   I noticed a couple of typos:
+) In guideline 1, there is a 'tha' which should be 'that'.
+) In request 3, there is a 'their' which should be 'there'.

**Comment by dblodgett-usgs** [comment]
typos corrected.

**Comment by davidhassell** [comment]
Hello,

Three weeks have passed with no further discussion, so I think we can accept this issue - i.e. the new CONTRIBUTING.md file.

Very many thanks to everyone who gave their time and git experiences to this discussion on the mailing list, at the CF meeting in June and here on the github issue.

David

**Comment by dblodgett-usgs** [comment]
Wonderful, thanks for working through this, everyone! This is a big step for the community. When we merge #137, we should also merge and deploy https://github.com/cf-convention/cf-convention.github.io/pull/62.

**Comment by JonathanGregory** [comment]
Dear all

That's very good. Thanks for proposing it, Dave.

David, would you be able to consider how and where to incorporate this in the governance page http://cfconventions.org/governance.html? Similarly, there is issue which Dave refers to about defects. How does these relate to trac ticket 160 and have we decided to stop using trac for new tickets? We need some clarity, I would say, so that proposers of changes know how to proceed.

Best wishes

Jonathan

**Comment by dblodgett-usgs** [comment]
Now that #137 is merged, I think we also need to merge: https://github.com/cf-convention/cf-convention.github.io/pull/62 right? Once that's done, I think this issue should be closed.

**Comment by davidhassell** [comment]
Absolutely right. It will be merged very soon ....

**Comment by ChrisBarker-NOAA** [comment]
Can we close this now? not good to have lingering issues....

I've started #150 -- where we can hash out the issues not addressed in the CONTRIBUTING.md doc

**Comment by JonathanGregory** [comment]
Dear Chris et al.
David Hassell has now got permissions to merge pull requests. I believe this will be done soon, now he is able to do it (since a couple of days ago).
Cheers
Jonathan

**Comment by davidhassell** [comment]
Just got my access permissions sorted out. https://github.com/cf-convention/cf-convention.github.io/pull/62 has been merged. Thanks

**Comment by dblodgett-usgs** [comment]
With @davidhassell's merge, I'm going to close this as complete.
